libclamav/ole2_extract.c

```diff
@@ -1756,6 +2024,423 @@ ole2_read_header(int fd, ole2_header_t *hdr)
+
+    ret = 0;
+done:
+    FREE(buffer);
+
+    if (0 == ret) {
+        memcpy(key, doubleSha, keyLength);
+    }
+
+    return ret;
+}
+
+static void aes_128ecb_decrypt(const unsigned char *in, size_t length, unsigned char *out, const uint8_t *key, unsigned key_n)
+{
+
+    unsigned long rk[RKLENGTH(128)];
+    int nrounds;
+    size_t i;
+
+    nrounds = rijndaelSetupDecrypt(rk, (const unsigned char *)key, key_n * 8);
+    if (!nrounds) {
+        cli_errmsg("ole2: nrounds NULL\n");
+    } else {
+        for (i = 0; i < length; i += 16) {
+            rijndaelDecrypt(rk, nrounds, &(in[i]), &(out[i]));
+        }
+    }
+}
+
+/*
+ * Returns true if it is actually encrypted with the key.
+ *
+ * Details here
+ * https://docs.microsoft.com/en-us/openspecs/office_file_formats/ms-offcrypto/e5ad39b8-9bc1-4a19-bad3-44e6246d21e6
+ */
+static bool verify_key(uint8_t key[16], encryption_verifier_t *verifier)
+{
+#define AES_VERIFIER_HASH_LEN 32
+
+    bool bRet = false;
+    uint8_t sha[SHA1_HASH_SIZE];
+    const uint32_t encrypted_verifier_hash_len = AES_VERIFIER_HASH_LEN;
+    uint8_t decrypted[encrypted_verifier_hash_len];
+
+    aes_128ecb_decrypt(verifier->encrypted_verifier, sizeof(verifier->encrypted_verifier),
+                       decrypted, key, 16);
+
+    cl_sha1(decrypted, sizeof(verifier->encrypted_verifier), sha, NULL);
+
+    aes_128ecb_decrypt(verifier->encrypted_verifier_hash, verifier->verifier_hash_size,
+                       decrypted, key, 16);
+
+    bRet = (0 == memcmp(sha, decrypted, verifier->verifier_hash_size));
+
+    return bRet;
+}
+
+//https://docs.microsoft.com/en-us/openspecs/office_file_formats/ms-offcrypto/dca653b5-b93b-48df-8e1e-0fb9e1c83b0f
+//https://learn.microsoft.com/en-us/openspecs/office_file_formats/ms-offcrypto/2895eba1-acb1-4624-9bde-2cdad3fea015
+static bool initialize_encryption_key(cli_ctx * ctx, const encryption_info_stream_standard_t *headerPtr,
+                                      encryption_key_t *encryptionKey)
+{
+
+    bool bRet  = false;
+    size_t idx = 0;
+    encryption_key_t key;
+    encryption_verifier_t ev;
+
+    memset(encryptionKey, 0, sizeof(encryption_key_t));
+    memset(&key, 0, sizeof(encryption_key_t));
+
+    cli_dbgmsg("Major Version   = 0x%x\n", headerPtr->version_major);
+    cli_dbgmsg( "Minor Version   = 0x%x\n", headerPtr->version_minor);
+    cli_dbgmsg( "Flags           = 0x%x\n", headerPtr->flags);
+
+    /*Bit 0 and 1 must be 0*/
+    if (1 & headerPtr->flags) {
+        cli_warnmsg("ole2: Invalid first bit, must be 0\n");
+        goto done;
+    }
+
+    if ((1 << 1) & headerPtr->flags) {
+        cli_warnmsg("ole2: Invalid second bit, must be 0\n");
+        goto done;
+    }
+
+#define SE_HEADER_FCRYPTOAPI (1 << 2)
+#define SE_HEADER_FEXTERNAL (1 << 4)
+#define SE_HEADER_FDOCPROPS (1 << 3)
+
+    //https://docs.microsoft.com/en-us/openspecs/office_file_formats/ms-offcrypto/200a3d61-1ab4-4402-ae11-0290b28ab9cb
```

Comment:
This link is also provided above where `flags` is defined, so it can probably be removed here.